docs/source/cuda/examples.rst

```diff
@@ -3,6 +3,479 @@
+   def solve_heat_equation(data, size, timesteps, k):
+      i = cuda.grid(1)
+
+      for step in range(timesteps):
+         # get the current temperature associated with this segment
+         curr_temp = data[i]
+         
+         # apply formula from finite difference equation
+         if i == 0:
+               # Left wall is held at T = 0
+               next_temp = curr_temp + k * (data[i + 1] - (2 * curr_temp))
+         elif i == size - 1:
+               # Right wall is held at T = 0
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp))
+         else:
+               next_temp = curr_temp + k * (data[i - 1] - (2 * curr_temp) + data[i + 1])
+         data[i] = next_temp
+         
+         # wait for every thread to write before moving on
+         cuda.syncthreads()
+
+Calling the kernel:
+
+.. code-block:: python
+
+   solve_heat_equation.forall(len(data))(data_gpu, len(data_gpu), 10000, 0.25)
+
+Since the kernel reused the original data vector to work on the problem, copying it back
+to the host yields the results:
+
+.. code-block:: python
+
+   print(data_gpu.copy_to_host())
+
+This prints
+
+.. code-block:: none
+
+   array([ 0.54632855,  1.09212753,  1.6368679 ,  2.18002166,  2.72106236,
+         3.25946564,  3.79470973,  4.32627596,  4.85364931,  5.37631887,
+         5.89377839,  6.40552673,  6.91106841,  7.409914  ,  7.90158064,
+         8.38559243,  8.86148082,  9.32878491,  9.78705184, 10.23583701,
+         10.67470447, 11.10322728, 11.52098821, 11.92758065, 12.32260997,
+         12.7056948 , 13.07646632, 13.43456455, 13.7796327 , 14.11129136,
+         14.42908512, 14.73343975, 15.0248911 , 15.30232574, 15.56474082,
+         15.81201588, 16.04395606, 16.26034615, 16.46097692, 16.64565132,
+         16.81418871, 16.9664262 , 17.10221779, 17.22143367, 17.32395972,
+         17.40969746, 17.47856409, 17.53049271, 17.56543251, 17.58334895,
+         17.58422393, 17.56805586, 17.53485974, 17.48466713, 17.41752618,
+         17.33350153, 17.23267429, 17.1151419 , 16.98101808, 16.83043268,
+         16.66353154, 16.48047636, 16.2814445 , 16.06662887, 15.83623766,
+         15.5904942 , 15.32963669, 15.053918  , 14.76360541, 14.45898035,
+         14.14033813, 13.80798763, 13.46225106, 13.10346357, 12.73197298,
+         12.34813942, 11.95233499, 11.5449434 , 11.12635958, 10.69698932,
+         10.25724886,  9.80756451,  9.3483722 ,  8.8801171 ,  8.40325314,
+         7.91824262,  7.42555573,  6.9256701 ,  6.41907034,  5.90624758,
+         5.38769897,  4.86392722,  4.33544009,  3.80274994,  3.26637319,
+         2.72682981,  2.18464289,  1.64033805,  1.09444297,  0.54748687])
+
+Plotting this data with a graphing library shows an arc shape highest where
+the object was hot initially and gradually sloping down to zero towards the
+edges where the temperature is fixed at zero. In the limit of infinite time,
+the arc will flatten out completely.
+
+.. _cuda_reduction_shared:
+
+Shared Memory Reduction
```

Comment:
We should probably note that this as an example of how to use shared memory rather than the way in which reductions should actually be implemented using Numba - we could point to the `@cuda.reduce` decorator (https://numba.readthedocs.io/en/stable/cuda/reduction.html) and/or demonstrate its use, as it provides an easier interface for building a reduction and uses a slightly more sophisticated (hopefully more performant) implementation: https://github.com/numba/numba/blob/main/numba/cuda/kernels/reduction.py

Reply:
I will add a comment here, and I used the `cuda.reduce` technique later on in the monte-carlo example.